src/runtime_src/core/pcie/driver/linux/xocl/subdev/xmc.c

```diff
@@ -2249,6 +2249,75 @@ static void xmc_clk_scale_config(struct platform_device *pdev)
 static int xmc_dynamic_region_free(struct platform_device *pdev);
 static int xmc_dynamic_region_freeze(struct platform_device *pdev);
 
+/*
+ * flags can be 1: Grant access
+ *              0: Release access
+ */
+static int cmc_access_ops(struct platform_device *pdev, int flags)
+{
+	xdev_handle_t xdev = xocl_get_xdev(pdev);
+	u32 val, grant, ack;
+	int retry;
+	int err = 0;
+
+	if (flags == 1) {
+		/*
+		 * for grant access (flags = 1), we are looking for new
+		 * features, if no new features, skip the grant operation
+		 */
+#if 1
+		err = xocl_iores_read32(xdev, XOCL_SUBDEV_LEVEL_URP,
+		    IORES_CLK_THROT, 0xb20, &val);
+		if (err) {
+			xocl_xdev_info(xdev, "No %s resource, skip.",
+			    NODE_CLK_THROT);
+			return 0;
+		}
+#else
+		/* workaround code if NODE_CLK_THROT isn't in xclbin yet */
+#endif
+		if (!(val & (1<<29))) {
+			xocl_xdev_info(xdev, "No new feature, skip grant CMC");
+			return 0;
+		}
+	}
+
+	grant = (u32)flags & 0x1;
+	err = xocl_iores_write32(xdev, XOCL_SUBDEV_LEVEL_BLD, IORES_CMC_MUTEX,
+	    0x0, grant);
+	if (err) {
+		xocl_xdev_info(xdev, "No %s resource, skip.", NODE_CMC_MUTEX);
+		return 0;
+	}
+
+	for (retry = 0; retry < 100; retry++) {
+		err = xocl_iores_read32(xdev, XOCL_SUBDEV_LEVEL_BLD,
+		    IORES_CMC_MUTEX, 0x8, &ack);
+		if (err) {
+			xocl_xdev_err(xdev, "Read ack from %s failed",
+			    NODE_CMC_MUTEX);
+			goto fail;
+		}
+
+		if ((grant & 0x1) == (ack & 0x1))
+			break;
+
+		msleep(100);
+	}
+
+	if ((grant & 0x1) != (ack & 0x1)) {
+		xocl_xdev_err(xdev,
+		    "bit 0 in Ack (0x%x) is not the same in grant (0x%x)",
+		    ack, grant);
+		err = -EBUSY;
+		goto fail;
+	}
+
+	xocl_xdev_info(xdev, "%s CMC succeeded.", flags ? "Grant" : "Release");
+fail:
+	return err;
+}
+
```

Comment:
xocl_iores_read32 and write may return -ENODEV if the resource is not found. Please add code to handle the case that cmc mutex does not exist.

Reply:
add -ENODEV checking